### ISSUE TITLE
Auto-reconnect documenation and samples are a bit confusing

### ISSUE THREAD

**Issue by vypsim** [dotnet/MQTTnet#1945]
Hi, the documentation for how to auto reconnect, when connection fails, is confusing. I'd appreciate it if I can get some clarification on the questions below.

First of all, the Wiki is referencing a method that no longer exists, which will lead developers to the samples below.

## [Reconnect_Using_Event Sample](https://github.com/dotnet/MQTTnet/blob/master/Samples/Client/Client_Connection_Samples.cs#L365)
The sample states:
> This approach has a risk of dead locks! Consider using the timer approach (see sample).

But the comment in [MqttClient.cs](https://github.com/dotnet/MQTTnet/blob/master/Source/MQTTnet/Client/MqttClient.cs#L592), where the Disconnected event is raised, says that it's wrapped in a `Task.Run(...)` to avoid dead-locks.

**Q1) Is running in a separate thread not enough and it may end up in a dead-lock anyway? or should the samples be updated and that comment removed to avoid confusion?**

## [Reconnect_Using_Timer Sample](https://github.com/dotnet/MQTTnet/blob/master/Samples/Client/Client_Connection_Samples.cs#L392)
The sample recommends calling `TryPingAsync` in a loop which seems incorrect after looking at the code in [MqttClient.cs](https://github.com/dotnet/MQTTnet/blob/master/Source/MQTTnet/Client/MqttClient.cs#L1035) since the client internally maintains a keep alive loop where it pings the server and updates the connection status (`IsConnected`) accordingly. And it does it in a way that's better than a simple loop with `TryPingAsync` because it takes into consideration whether packets have been exchanged succesfully and thus avoids the extra pings.
And more to this point, when looking at `ManagedMqttClient` which has a timer loop, it doesn't call `Ping` at all, but instead checks `IsConnected`

**Q2) Is `Reconnect_Using_Timer` sample misleading and inefficient or have I missed something?**

**Comment by chkr1011** [comment]
> Q1) Is running in a separate thread not enough and it may end up in a dead-lock anyway? or should the samples be updated and that comment removed to avoid confusion?

You are right but there are situations in tickets where a deadlock still happens. Also users are not using async sometimes and call .Result .Wait etc. instead. This is where this recommendation comes from.

> Q2) Is Reconnect_Using_Timer sample misleading and inefficient or have I missed something?

The ManagedClient is no longer developed because ist has several conceptual issues.
The sample also works when the keep alive interval is 0 and no ping message is sent internally.
Another reason is that the keep alive interval can be very high like 2 minutes etc. Then the user can call PingAsync to identify if the broker is available in this particular moment. Depending on your needs there is no need to call the PingAsnyc.
